### ISSUE TITLE
CMDER here Context Menu not working

### ISSUE THREAD

**Issue by TreeBranches** [cmderdev/cmder#2645]
Related; https://github.com/cmderdev/cmder/issues/1163

We're trying to deploy CMDER across our domain, and in configuration we've specified the user's home directory (C:\UsersLocal\$Username instead of C:\Users\$Username). According to that issue (and several other reported issues), specifying a directory location breaks the "Open Cmder Here" context menu.

I've seen various attempts at fixing this but none of them have proved to work for me, though I'm not sure if I'm doing it correctly. Things I've tried;

In .\vendor\init.bat I've added the following

```
if defined CMDER_START (
    cd /d "%CMDER_START%"
)
```

The context menu continues to not work.

In .\vendor\ConEmu.xml.default in the CMD1 line for cmd:Cmder task I've changed it from 

```cmd /k ""%ConEmuDir%\..\init.bat" -new_console:d:C:\UsersLocal\%USERNAME%```

to

```cmd /k ""%ConEmuDir%\..\init.bat" -new_console:d:%CMDER_START%```

And changing the Registry Key values to `"C:\Program Files (x86)\cmder\Cmder.exe" /START "%V"` in the following registry keys:

`Computer\HKEY_CLASSES_ROOT\Directory\shell\Cmder\command\(Default)`
`Computer\HKEY_CLASSES_ROOT\Directory\Background\shell\Cmder\command\(Default)`

This causes CMDER to open with an error (Error code 267) through both the context menu and the start menu, though Click Yes to open a new Terminal Window will feed the wanted directory through - this isn't a good solution really.

For good measure I have also specified System Environment Variables;
CMDER_ROOT = C:\Progra~2\cmder\
ConEmuDir = C:\Progra~2\cmder\vendor\conemu-maximus5
PATH = C:\Progra~2\cmder\

Any other suggestions as to where I'm going wrong?

**Comment by chrisant996** [comment]
I see that a short pathname is being used for "c:\Program Files (x86)".  I just want to make sure that you're aware that short pathnames can't be hard-coded, because they can be different on different specific computers.  If the short pathname is being hard-coded, then that may contribute to malfunctions.

Some other questions:
- Where is "UsersLocal" coming from?  I've never seen that as a directory name on Windows before.
- Why is "Users$Username" or "UsersLocal$Username" syntax being used?  That looks like a malformed path (missing `\`) and also UNIX style environment variable naming (`$var`) being used on Windows (which uses `%var%`).

**Comment by daxgames** [comment]
Why are you hard coding anything?   You should never edit `init.bat`, why are you wditing `init.bat`?

No need to set anything in the registry, no need to set env variables.

See: [This](
https://github.com/cmderdev/cmder/issues/2640#issuecomment-999557571)

Yo can add to the installer in the link referenced above running rhe command `"[path to cmder folder]\cmder.exe /register"

Hope this helps.

**Comment by daxgames** [comment]
I am sorry I do not fully even understand the issue.  The context menu works fine if using the `/register` command and Cmder is not in `C:\Program Files`.

Cmder was never meant to be installed into `C:\Program Files` since all user config is stored by default in the Cmder `config` folder.  We added the `/c [config_path]` so users could do a shared install in a central location like `C:\Program Files\Cmder` but FULL functionality was never promised.

If I put Cmder in `C:\Program Files\Cmder` I am unable to get a functioning Shell context menu.  I get the below:

![image](https://user-images.githubusercontent.com/7318053/149831372-e321d23b-6bcd-4153-b7f8-7daf99028195.png)

I assume the above is what you mean by 'CMDER here Context Menu not working'.

**Comment by TreeBranches** [comment]
The issue you're experiencing is due to folder permissions - these are issues that I have resolved already through the MSI installation.

The issue that I'm trying to resolve specifically is the Right Click > Open Cmder Here not actually opening the folder that's being right clicked, be it in the white space or the folder in question. Instead, it is only opening Cmder at the starting location that I specified in ConEmu.xml.default

[Here is a link to the MSI which we deploy](https://drive.google.com/file/d/1tYQCHMzPEYO_UtKAcL3xxNNQQ9hKewca/view?usp=sharing), perhaps you can see where in the structure/configuration I'm going wrong in terms of PATH, configuration file, registry key, or maybe even a folder or file permission I'm not aware of. The only thing you'll need to do manually is make a C:\UsersLocal\*your.username.here* folder, so that the default start location works and doesn't give you an error when launching.

**Comment by daxgames** [comment]
If you specify a folder in the taks that is the one that will be opened.  Remove that an the context menu should work.

**Comment by TreeBranches** [comment]
> If you specify a folder in the task that is the one that will be opened. Remove that an the context menu should work.

Is this not required in order to specify the starting directory as `C:\UsersLocal\$username` as opposed to `C:\Users\$username` though?

**Comment by daxgames** [comment]
No it is not.

**Comment by TreeBranches** [comment]
I see.

OK, for now, I've removed all pointers to the UsersLocal folder. The result is that CMDER now starts in C:\Program Files (x86)\cmder but at least the context menu is working.

Would be great if this could be fixed so that it was possible to have both a default directory and a context menu in place.

**Comment by chrisant996** [comment]
Consider using separate tasks, instead of using only one.  It looks like you can make a separate task that is only for the context menu (see the Integration node in Settings), and in that task do not override the startup folder.

**Comment by daxgames** [comment]
@TreeBranches as I have said a default directory is already possible if you create a shortcut with a 'Start in Folder' specified.

Run `cmder.exe [folder path]` to open a shell in that path to prove it.

Setting a default folder as you are trying is not in our control.  That is a function of ConEmu which we do not develop.

You could use separate tasks as @chrisant996  suggested but that just adds even more unnecessary complexity.

The only real reason to put Cmder in `C:\Program Files (x86)\cmder` and use `/c [user config path` vs. in `[user home folder]\cmder` is if you are installing cmder on a multi user system to keep user configs private.

All would work fine if you would just install Cmder where we recommend in `[user home folder]\cmder` and have your installer execute `[user home folder]\cmder\cmder.exe /register`.  Gives you user private config files, default open folder `[user home folder]\cmder`, and a shell extension that works properly.  No special file system permissions, no custom config.  The only other thing you might want to add is a Start Menu shortcut with a 'Start in Folder' specified set to `[user home folder]`.

You are making it harder than it needs to be just to put Cmder in a folder called `C:\Program Files (x86)` even though Cmder is NOT supposed to go there.

**Comment by TreeBranches** [comment]
> Consider using separate tasks, instead of using only one. It looks like you can make a separate task that is only for the context menu (see the Integration node in Settings), and in that task do not override the startup folder.

I'm going to have a more thorough look at this later today and post an update, but I don't see any documentation on how to correctly set this up. It sounds like the solution to all this though. Do you have a link with more information?

~

> The only real reason to put Cmder in `C:\Program Files (x86)\cmder` and use `/c [user config path` vs. in `[user home folder]\cmder` is if you are installing cmder on a multi user system to keep user configs private.

This is exactly what we're trying to do. I heavily implied this when my first line in this post us wanting to deploy this in a domain environment.

> All would work fine if you would just install Cmder where we recommend in [user home folder]\cmder and have your installer execute [user home folder]\cmder\cmder.exe /register.

Software deployed in a domain gets deployed as NTAUTH\System, so any script I execute during install will only apply to that user and inaccessible to anyone else. It could perhaps be organised as a user sign in script potentially I suppose, though I'm concerned about sign in times with that, and is more tricky to manage updates with - generally best to keep things like that to a minimum. If cmder was added to the Windows Store, we'd be able to work with that through a deployment script or the Windows Store for Business, but unfortunately cmder isn't on the Windows Store.

In short, installing software to AppData is the worst thing possible in a domain environment and at best just creates data duplication. I understand that certain elements are outside of cmder's scope and more ConEmu related.

**Comment by daxgames** [comment]
@TreeBranches did not realize toy were using windows software deployment to deploy the software so the user folder based install would jot work.

Since you are editing the conemu.xml already you could.

1. Add a default start folder to the Cmder default start task like you did before.
2. Add a new task with no default start folder called `Cmder-shell-menu` then modify the registry entries for the context menu integration to include `/task Cmder-shell-menu`.

You would still be storing user config in `C:\Program Files (x86)\Cmder\` so not ideal but it must be OK with you.

If you know a C++ developer that might be able to fix the `/c [user config path]` we do accept external contributions.  I just do not know how to fix it.

**Comment by daxgames** [comment]
@TreeBranches did you get a satisfactory solution to your issues?

**Comment by daxgames** [comment]
Closing no response from OP
